core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -429,20 +466,84 @@ public HashDistributionTaskSource(
+                        // replicated exchange source will be added in postprocessTasks below
                         .build();
                 HostAddress host = partitionToNodeMap.get(partition);
                 Set<HostAddress> hostRequirement = host == null ? ImmutableSet.of() : ImmutableSet.of(host);
-                result.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement)));
+                partitionTasks.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement)));
             }
 
+            List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+
             finished = true;
-            return result.build();
+            return result;
+        }
+
+        private List<TaskDescriptor> postprocessTasks(List<TaskDescriptor> tasks)
+        {
+            ListMultimap<NodeRequirements, TaskDescriptor> taskGroups = groupCompatibleTasks(tasks);
+            ImmutableList.Builder<TaskDescriptor> joinedTasks = ImmutableList.builder();
+            long replicatedExchangeSourcesSize = replicatedExchangeSourceHandles.values().stream().mapToLong(this::sourceHandleSize).sum();
+            int taskPartitionId = 0;
+            for (Map.Entry<NodeRequirements, Collection<TaskDescriptor>> taskGroup : taskGroups.asMap().entrySet()) {
+                NodeRequirements groupNodeRequirements = taskGroup.getKey();
+                Collection<TaskDescriptor> groupTasks = taskGroup.getValue();
+
+                ImmutableListMultimap.Builder<PlanNodeId, Split> splits = ImmutableListMultimap.builder();
+                ImmutableListMultimap.Builder<PlanNodeId, ExchangeSourceHandle> exchangeSources = ImmutableListMultimap.builder();
+                long splitsWeight = 0;
+                long exchangeSourcesSize = 0;
+
+                for (TaskDescriptor task : groupTasks) {
+                    ListMultimap<PlanNodeId, Split> taskSplits = task.getSplits();
+                    ListMultimap<PlanNodeId, ExchangeSourceHandle> taskExchangeSources = task.getExchangeSourceHandles();
+                    long taskSplitWeight = taskSplits.values().stream().mapToLong(split -> split.getSplitWeight().getRawValue()).sum();
+                    long taskExchangeSourcesSize = taskExchangeSources.values().stream().mapToLong(this::sourceHandleSize).sum();
+
+                    if ((splitsWeight > 0 || exchangeSourcesSize > 0)
+                            && ((splitsWeight + taskSplitWeight) > targetPartitionSplitWeight || (exchangeSourcesSize + taskExchangeSourcesSize + replicatedExchangeSourcesSize) > targetPartitionSourceSizeInBytes)) {
+                        exchangeSources.putAll(replicatedExchangeSourceHandles); // add replicated exchanges
+                        joinedTasks.add(new TaskDescriptor(taskPartitionId++, splits.build(), exchangeSources.build(), groupNodeRequirements));
```

Comment:
QQ: does task execution automatically takes care of processing the splits and partitions corresponding to the same hash?

Reply:
Execution is dumb. But each of `TaskDescriptor`s that is passed to `postprocessTasks` handles a single partition (splits read data from table buckets that match exchange partition).
In `postprocessTasks` I merge some `TaskDescriptors` if they are too small.